### PR TITLE
Internal PlusCal parser, misc. changes and fixes

### PR DESCRIPTION
**Pull request by fhackett**
This set of changes should probably have been several PRs - sorry for the huge diff. The general theme is parser rework and minor fixes.

Here is a list of changes:
- implemented PlusCal parser under pgo.parser.PlusCalParser. Both C- and P- syntax are supported, though some more testing will be needed as things progress.
- completely reworked the parsing engine for the TLA+ and now PlusCal parsers, it's faster and nicer to work with. We also have general backtracking, so can now correctly parse (then scratch our heads over, since it's meaningless) this: `WF_a(b)(c)`.
- implemented some minor AST renamings and changes to keep things consistent as we expand the compiler
- removed the TLC codebase/JAR as we no longer need them
- ran some code sanitization steps via IntelliJ in order to remove dead imports, etc
- fixed some parsing bugs that were missed in my previous PR

### CODE REVIEW

**Comment by fhackett** [comment]
@mrordinaire It seems this PR also resolves issue #84 somewhere in its huge pile of changes. I have added a test for that.

File: build.xml [hunk 1]
```diff
@@ -26,7 +26,6 @@
   <path id="pgo.depjars.classpath">
     <pathelement location="lib/hamcrest-all-1.3.jar"/>
     <pathelement location="lib/junit-4.12.jar"/>
-    <pathelement location="lib/tlatools.jar"/>
```
Comment (mrordinaire) [inline build.xml]: Hooray!

File: src/pgo/PGoNetOptions.java [hunk 2]
```diff
@@ -6,7 +6,7 @@
```
Comment (mrordinaire) [inline src/pgo/PGoNetOptions.java]: I suspect these are due to some autorefactoring tool. Please revert them.

File: src/pgo/PGoOptions.java [hunk 3]
```diff
@@ -1,15 +1,16 @@
...
-import java.io.IOException;
-import java.nio.file.Files;
-import java.nio.file.Paths;
+ import java.io.IOException;
+ import java.nio.file.Files;
+ import java.nio.file.Paths;
 
 public class PGoOptions {
-	@Option(value = "-h Print usage information", aliases = { "-help" })
+	@Option(value = "-h PlusCalPrint usage information", aliases = { "-help" })
```
Comment (mrordinaire) [inline src/pgo/PGoOptions.java]: Please revert this change.

File: src/pgo/PGoOptions.java [hunk 4]
```diff
@@ -18,7 +19,7 @@
 	/**
 	 * Be verbose, print extra detailed information. Sets the log level to FINE.
 	 */
-	@Option(value = "-v Print detailed information during execution ", aliases = { "-verbose" })
+	@Option(value = "-v PlusCalPrint detailed information during execution ", aliases = { "-verbose" })
```
Comment (mrordinaire) [inline src/pgo/PGoOptions.java]: As well as this change

File: src/pgo/formatters/IssueFormattingVisitor.java [hunk 5]
```diff
@@ -82,10 +89,15 @@ public Void visit(IOErrorIssue ioErrorIssue) throws IOException {
...
 
 	@Override
-	public Void visit(TLAParserIssue tlaParserIssue) throws IOException {
-		out.write("could not parse TLA: ");
+	public Void visit(ParsingIssue parsingIssue) throws IOException {
+		Map.Entry<SourceLocation, Set<ParseFailure>> lastEntry = parsingIssue.getError().lastEntry();
+		out.write("could not parse "+parsingIssue.getLanguage()+" at "+lastEntry.getKey()+": ");
 		try(IndentingWriter.Indent ignored = out.indent()){
-			out.write(tlaParserIssue.getError().toString());
+			Set<ParseFailure> lastFailures = lastEntry.getValue();
```
Comment (mrordinaire) [inline src/pgo/formatters/IssueFormattingVisitor.java]: Why choose a set? This places the burden of implementing sensible logic for equality and hashing on the next guy.

File: src/pgo/model/type/PGoTypeGenerator.java [hunk 6]
```diff
@@ -1,11 +1,12 @@
...
 
-import java.util.*;
+import java.util.HashMap;
+import java.util.HashSet;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Supplier;
 
 /**
- * Generates fresh PGoTypeVariables and infers types from Go type names.
+ * Generates fresh PGoTypeVariables and infers types from GoRoutineStatement type names.
```
Comment (mrordinaire) [inline src/pgo/model/type/PGoTypeGenerator.java]: Please revert this.

File: test/pgo/ExpressionCodeGenRunTest.java [hunk 7]
```diff
@@ -268,7 +270,7 @@ public void test() throws IOException {
 				System.out.println("source: "+line);
 			});
 
-			// try to run the compiled Go code, check that it prints the right thing
+			// try to run the compiled GoRoutineStatement code, check that it prints the right thing
```
Comment (mrordinaire) [inline test/pgo/ExpressionCodeGenRunTest.java]: Please revert this

File: test/pgo/IntegrationTestingUtils.java [hunk 8]
```diff
@@ -28,7 +28,7 @@ public static void expungeFile(File file) {
 	}
 	
 	public static void testRunGoCode(Path codePath, List<String> expected) throws IOException {
-		// try to run the compiled Go code, check that it prints the right thing
+		// try to run the compiled GoRoutineStatement code, check that it prints the right thing
```
Comment (mrordinaire) [inline test/pgo/IntegrationTestingUtils.java]: Please revert this

**Comment by mrordinaire** [comment]
@fhackett, could you please add more tests for PlusCal parsing?

File: src/pgo/formatters/IssueFormattingVisitor.java (hunk 5 above)
Comment (fhackett) [reply src/pgo/formatters/IssueFormattingVisitor.java]: The set was chosen since it is common for there to be many instances of the same parse failure in the same place - it forces deduplication of the failures on store. I do know it's not very elegant (unordered, for example), but I'm not sure about what it means to place a burden on the next guy... I've already implemented hashing and equality for all failure types?

**Comment by fhackett** [comment]
Rebased onto latest master - I'm actually starting to think that it may be easier for others to rebase/merge onto this at this point. Any thoughts?

**Comment by fhackett** [comment]
(the changes are relatively small, you just have to rename the AST nodes you reference)

**Comment by adam-tg** [comment]
I think if the changes are small it makes more sense to push the work to the smaller changes. I'm happy to do the renaming if you want.

**Comment by mrordinaire** [comment]
I'll rebase my changes onto this

**Comment by mrordinaire** [comment]
Let's merge this and fix bugs as we find them
